# GES_2020
Code (data analysis and model simulations) for Genetic (G) / Epigenetic (E) / Stochastic (S)  paper (2020). 

DrugResponse folder includes drug-response count data (csv), annotated by files in
the platemap folder. Two scripts are included, functions from the previously
published diprate R package (functionsDRC.R; Harris et al., Nat Meth (2016))
and the script to reproduce drug response figures in the paper (DrugResponse.R).

clonal Fractional Proliferation (cFP) folder includes single-cell data for each
sample (one csv = one 384-well plate, each well with single-colony growth
rates). cFP.R analysis script plots figures in text associated with drug-induced
proliferation (DIP) rate distributions. PopD_trajectories.RData includes data from the
DrugResponse folder for a comparison plot. Various trajectories*, distributions*, and
pvalues* files are simulated data (see Simulations folder below), which are directly
compared to the experimental cFP data.

Joint_functions includes plotting functions shared between the cFP and DrugResponse
folders.

Simulations folder includes the monoclonal and polyclonal growth model (M/PGM) and associated
parameter scans (.py files). Parameter scan data is provided in the form of a pickled dataframe
(.pkl files). Two scripts (convertPKLtoCSV*.py) convert the dataframes to csv files plotted in 
the plotParameterScan.R script. Plotting of example model simulation data happens in the
cFP folder (see above). 

Whole exome sequencing (WES) folder includes bash scripts (txt files) for the conversion of fastq 
files (see SRA accession number, below) to variant call format (vcf) and variant effect predictor
(vep) annotated datasets. Additionally, summarized metrics from the data (csv files), an .xlm 
file to identify the canonical ex19del mutation in all cell populations, and reference genome (rda
file). Code for analysis and figure generation are provided (WES.R).

Single-cell RNA sequencing (scRNAseq) folder includes three data folders, two scripts, 
and some resource files. The read_count folder (generated by 10X Genomics CellRanger) 
includes the gene expression library for all 8 cell populations, multiplexed and sequenced 
together. The HTO_count bash script (provided as .txt file) contains the script to generate 
the umi_count folder. The umi_count folder includes the hashtag oligonucleotide (HTO) library, 
created by CITE-seq Count function on the 'hashed' data. The HTO_identification folder provides 
additional information on the HTO library. scRNAseq.R creates figures from the paper, and findDEGs.R 
identifies lists of differentially expressed genes (DEGs) to be used in the GO folder (see below). 
The cell cycle genes file (.txt) is used for a cell cycle regression analysis in the scRNAseq.R script.

The gene ontology (GO) folder has a script that performs a GO, graph-structure based semantic 
similarity analysis on mutations and DEGs that define each cell population (semanticSimilarity.R), in 
order to find a genetic-to-epigenetic connection between the datasets. Two RData files included in 
the folder are calculated separately in the WES and scRNAseq folders, and correspond to IMPACT mutations
and differentially expressed genes, respectively.

RNA sequencing (RNAseq) folder includes a bash script (RNAseq_processing.txt) that generates a count matrix
from the fastq files and an R script that runs a simple clustering and visualization of the bulk RNAseq data 
for all cell populations.


Seqeuncing data can be accessed in the following databases:
	SRA (PRJNA631050, PRJNA632351)
	GEO (GSE150084)
